android/src/main/java/com/oblador/keychain/KeychainModule.java

```diff
@@ -97,59 +117,96 @@ public void setGenericPasswordForOptions(String service, String username, String
     }
 
     @ReactMethod
-    public void getGenericPasswordForOptions(String service, Promise promise) {
+    public void getGenericPasswordForOptions(String service, final Promise promise) {
+        final String serviceOrDefault = getDefaultServiceIfNull(service);
+        CipherStorage cipherStorage = null;
         try {
-            service = getDefaultServiceIfNull(service);
-
-            CipherStorage currentCipherStorage = getCipherStorageForCurrentAPILevel();
-
-            ResultSet resultSet = prefsStorage.getEncryptedEntry(service);
+            ResultSet resultSet = prefsStorage.getEncryptedEntry(serviceOrDefault);
             if (resultSet == null) {
-                Log.e(KEYCHAIN_MODULE, "No entry found for service: " + service);
+                Log.e(KEYCHAIN_MODULE, "No entry found for service: " + serviceOrDefault);
                 promise.resolve(false);
                 return;
             }
 
-            final DecryptionResult decryptionResult = decryptCredentials(service, currentCipherStorage, resultSet);
-
-            WritableMap credentials = Arguments.createMap();
-
-            credentials.putString("service", service);
-            credentials.putString("username", decryptionResult.username);
-            credentials.putString("password", decryptionResult.password);
+            // Android < M will throw an exception as biometry is not supported.
+            CipherStorage biometryCipherStorage = null;
+            try {
+                biometryCipherStorage = getCipherStorageForCurrentAPILevel(true);
+            } catch(Exception e) { }
+            final CipherStorage nonBiometryCipherStorage = getCipherStorageForCurrentAPILevel(false);
+            if (biometryCipherStorage != null && resultSet.cipherStorageName.equals(biometryCipherStorage.getCipherStorageName())) {
+                cipherStorage = biometryCipherStorage;
+            } else if (nonBiometryCipherStorage != null && resultSet.cipherStorageName.equals(nonBiometryCipherStorage.getCipherStorageName())) {
+                cipherStorage = nonBiometryCipherStorage;
+            }
 
-            promise.resolve(credentials);
-        } catch (KeyStoreAccessException e) {
-            Log.e(KEYCHAIN_MODULE, e.getMessage());
-            promise.reject(E_KEYSTORE_ACCESS_ERROR, e);
+            final CipherStorage currentCipherStorage = cipherStorage;
+            if (currentCipherStorage != null) {
```

Comment:
the bodies of the `if` and `else` blocks are quite large, could you extract them to separate methods with good names?